src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceRunner.java

```diff
@@ -118,161 +106,50 @@ public void runJob(final ScheduledJobParameter jobParameter, final JobExecutionC
-        try {
-            List<String> deletedIndices = new ArrayList<>();
-            for (String index : parameter.getIndices()) {
-                if (index.equals(parameter.currentIndexName())) {
-                    continue;
+            lockService.acquireLock(jobParameter, context, ActionListener.wrap(lock -> {
+                if (lock == null) {
+                    return;
                 }
-
-                if (!clusterService.state().metadata().hasIndex(index)) {
-                    deletedIndices.add(index);
-                    continue;
-                }
-
                 try {
-                    if (client.admin()
-                        .indices()
-                        .prepareDelete(index)
-                        .setIndicesOptions(IndicesOptions.LENIENT_EXPAND_OPEN)
-                        .execute()
-                        .actionGet(timeout)
-                        .isAcknowledged()) {
-                        deletedIndices.add(index);
-                    } else {
-                        log.error("Failed to delete an index [{}]", index);
-                    }
-                } catch (Exception e) {
-                    log.error("Failed to delete an index [{}]", index, e);
+                    updateDatasource(jobParameter);
+                } finally {
+                    lockService.release(
+                        lock,
+                        ActionListener.wrap(released -> {}, exception -> { log.error("Failed to release lock [{}]", lock, exception); })
+                    );
                 }
-            }
-            if (!deletedIndices.isEmpty()) {
-                parameter.getIndices().removeAll(deletedIndices);
-                DatasourceHelper.updateDatasource(client, parameter, timeout);
-            }
-        } catch (Exception e) {
-            log.error("Failed to delete old indices for {}", parameter.getId(), e);
-        }
+            }, exception -> { log.error("Failed to acquire lock for job [{}]", jobParameter.getName(), exception); }));
+        };
     }
 
-    /**
-     * Update GeoIP data internal
-     *
-     * @param jobParameter
-     * @throws Exception
-     */
-    private void updateDatasource(final Datasource jobParameter) throws Exception {
-        if (!DatasourceState.AVAILABLE.equals(jobParameter.getState())) {
-            log.error("Invalid datasource state. Expecting {} but received {}", DatasourceState.AVAILABLE, jobParameter.getState());
-            jobParameter.disable();
-            jobParameter.getUpdateStats().setLastFailedAt(Instant.now());
-            DatasourceHelper.updateDatasource(client, jobParameter, timeout);
+    @VisibleForTesting
+    protected void updateDatasource(final ScheduledJobParameter jobParameter) throws IOException {
+        Datasource datasource = datasourceFacade.getDatasource(jobParameter.getName());
+        if (datasource == null) {
+            log.info("Datasource[{}] does not exist", jobParameter.getName());
```

Comment:
Should this throw an exception?

Reply:
It can happen in normal use case. Delete request can be received while update task being waiting on queue for other update task to complete. When its turn come, it found that the datasource is deleted already so it should stop the update process.

Reply:
Can you update it in doc?

Reply:
Done